### ISSUE TITLE
New "Expand declarative macros" engine results in false positive errors

### ISSUE THREAD

**Issue by ewirch** [intellij-rust/intellij-rust#6843]
## Environment

* **IntelliJ Rust plugin version:** 0.3.140.3644-203
* **Rust toolchain version:** 1.51.0-nightly (04caa632d 2021-01-30) x86_64-unknown-linux-gnu
* **IDE name and version:** IntelliJ IDEA 2021.1 EAP Ultimate Edition (IU-211.5787.15)
* **Operating system:** Linux 5.10.15-1-manjaro
* **Macro expansion engine:** new
* **Name resolution engine:** old

## Problem description
Code from [Rust By Example](https://doc.rust-lang.org/stable/rust-by-example/error/multiple_error_types/option_result.html) gives wrong errors:
```rust
use std::num::ParseIntError;

fn double_first(vec: Vec<&str>) -> Result<Option<i32>, ParseIntError> {
    let opt = vec.first().map(|first| {
        first.parse::<i32>().map(|n| 2 * n)
    });

    opt.map_or(Ok(None), |r| r.map(Some))
}

fn main() {
    let numbers = vec!["42", "93", "18"];
    let empty = vec![];
    let strings = vec!["tofu", "93", "18"];

    println!("The first doubled is {:?}", double_first(numbers));
    println!("The first doubled is {:?}", double_first(empty));
    println!("The first doubled is {:?}", double_first(strings));
}
```

The identifier `n` in the fifth line shows error: 
> mismatched types [E0308] expected `Duration`, found `i32`

The `println!` lines show error:
> `Result<Option<i32>, ParseIntError>` doesn't implement `Debug` (required by {:?})

Both errors don't show up, when code is pasted using "Use old engine".

At the same time, this exception is logged:
```
2021-02-14 06:26:20,456 [3503246]  ERROR - intellij.openapi.progress.Task - Cannot create class org.rust.lang.core.macros.MacroExpansionSharedCache (classloader=PluginClassLoader(plugin=PluginDescriptor(name=Rust, id=org.rust.lang, descriptorPath=plugin.xml, path=~/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.5787.15.plugins/intellij-rust, version=0.3.140.3644-203, package=null), packagePrefix=null, instanceId=152, state=active)) 
com.intellij.diagnostic.PluginException: Cannot create class org.rust.lang.core.macros.MacroExpansionSharedCache (classloader=PluginClassLoader(plugin=PluginDescriptor(name=Rust, id=org.rust.lang, descriptorPath=plugin.xml, path=~/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.5787.15.plugins/intellij-rust, version=0.3.140.3644-203, package=null), packagePrefix=null, instanceId=152, state=active))
	at com.intellij.ide.plugins.PluginManagerCore.createPluginException(PluginManagerCore.java:300)
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:12)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:81)
	at com.intellij.serviceContainer.ComponentManagerImpl.instantiateClass(ComponentManagerImpl.kt:760)
	at com.intellij.serviceContainer.ComponentManagerImpl.createLightService(ComponentManagerImpl.kt:699)
	at com.intellij.serviceContainer.ComponentManagerImpl.access$createLightService(ComponentManagerImpl.kt:55)
	at com.intellij.serviceContainer.ComponentManagerImpl$getOrCreateLightService$1.run(ComponentManagerImpl.kt:596)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:688)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:634)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:63)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:218)
	at com.intellij.serviceContainer.ComponentManagerImpl.getOrCreateLightService(ComponentManagerImpl.kt:595)
	at com.intellij.serviceContainer.ComponentManagerImpl.getLightService$intellij_platform_serviceContainer(ComponentManagerImpl.kt:566)
	at com.intellij.serviceContainer.ComponentManagerImpl.doGetService(ComponentManagerImpl.kt:478)
	at com.intellij.serviceContainer.ComponentManagerImpl.getService(ComponentManagerImpl.kt:471)
	at org.rust.lang.core.macros.MacroExpansionSharedCache$Companion.getInstance(MacroExpansionSharedCache.kt:361)
	at org.rust.lang.core.macros.MacroExpansionTaskBase.run(MacroExpansionTask.kt:148)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:997)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:497)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:228)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:688)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:634)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:63)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:228)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.GeneratedConstructorAccessor136.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.intellij.serviceContainer.ComponentManagerImpl.instantiateClass(ComponentManagerImpl.kt:722)
	... 30 more
Caused by: java.lang.NoSuchMethodError: 'void com.intellij.psi.stubs.SerializedStubTreeDataExternalizer.<init>(com.intellij.psi.stubs.SerializationManagerEx, com.intellij.psi.stubs.StubForwardIndexExternalizer)'
	at org.rust.lang.core.macros.CompatUtilsKt.newSerializedStubTreeDataExternalizer(compatUtils.kt:24)
	at org.rust.lang.core.macros.PersistentCacheData$Companion.tryCreate(MacroExpansionSharedCache.kt:263)
	at org.rust.lang.core.macros.MacroExpansionSharedCache.tryCreateData(MacroExpansionSharedCache.kt:79)
	at org.rust.lang.core.macros.MacroExpansionSharedCache.<init>(MacroExpansionSharedCache.kt:61)
	... 34 more
```

## Steps to reproduce
Paste the above code in to editor, while "Use new engine" is selected for "Expand declarative macros".

**Comment by Undin** [comment]
Fixed by #6741
You can wait next stable release or try nightly plugin builds where this fix is already available. See https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly how to do it
